### PR TITLE
binaryen: 102 -> 105; emscripten: 3.0.0 -> 3.1.10

### PR DESCRIPTION
**Pull request by duxovni**
###### Description of changes

emscripten 3.1.1 and later include multiple fixes for nondeterministic build output; see emscripten-core/emscripten#15780 and emscripten-core/emscripten#15844.

I've combined the emscripten upgrade and the binaryen upgrade into one pull request, because emscripten breaks if either of them is upgraded without upgrading the other one (as in #159401).

Several packages with dependencies on emscripten/binaryen are broken; all of them were already broken on `master` and are unaffected by this change:
- emscriptenPackages.zlib (#170621)
- magnetophonDSP.ConstantDetuneChorus
- faustStk

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by willcohen** [comment]
Result of `nixpkgs-review pr 172741` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>open-music-kontrollers.mephisto</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>binaryen</li>
    <li>emscripten</li>
  </ul>
</details>

**Review by willcohen — APPROVED** [review]

**Comment by willcohen** [comment]
When testing a rebase of #172207 on top of this (to determine potential breakage), `emscriptenPackages.zlib` and `emscriptenPackages.json_c` still continue to build as noted on that PR, but the otherwise fixed `emscriptenPackages.libxml2` now fails with the following error:

```
$ nix-build $NIXPKGS -A emscriptenPackages.libxml2
this derivation will be built:
  /nix/store/8qid47b023i2vx7wv1q2ixbjk1a70p8m-emscripten-libxml2-2.9.14.drv
building '/nix/store/8qid47b023i2vx7wv1q2ixbjk1a70p8m-emscripten-libxml2-2.9.14.drv'...
unpacking sources
unpacking source archive /nix/store/x8swlr47ad8y0v15ssbq2c8x5abnn602-libxml2-2.9.14.tar.xz
source root is libxml2-2.9.14
setting SOURCE_DATE_EPOCH to timestamp 1651491766 of file libxml2-2.9.14/libxml2.spec
patching sources
applying patch /nix/store/frr6f2hq56j6b4wc0bsklaqskjsagqc0-utf8-xmlErrorFuncHandler.patch
patching file python/libxml.c
Hunk #1 succeeded at 1621 (offset 1 line).
Hunk #2 succeeded at 1638 (offset 1 line).
autoreconfPhase
autoreconfPhase not used...
configuring
configure: ./configure --prefix=/nix/store/11kprvkv7gs87l1sgl1hl1l76kw6pk61-emscripten-libxml2-2.9.14 --without-python
checking whether to enable maintainer-specific portions of Makefiles... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for a BSD-compatible install... /nix/store/5imadx3sbb1f2gqhxhw5rq5idwqj4mib-coreutils-9.0/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /nix/store/5imadx3sbb1f2gqhxhw5rq5idwqj4mib-coreutils-9.0/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking for gcc... /nix/store/flviblq5i53jgpzj8s5dpw3pdppxycpl-emscripten-3.1.10/bin/emcc
checking whether the C compiler works... no
configure: error: in `/build/libxml2-2.9.14':
configure: error: C compiler cannot create executables
See `config.log' for more details
emconfigure: error: './configure --prefix=/nix/store/11kprvkv7gs87l1sgl1hl1l76kw6pk61-emscripten-libxml2-2.9.14 --without-python' failed (returned 77)
error: builder for '/nix/store/8qid47b023i2vx7wv1q2ixbjk1a70p8m-emscripten-libxml2-2.9.14.drv' failed with exit code 1
```

**Comment by duxovni** [comment]
Hmm, interesting. Looks like the error relates to https://github.com/emscripten-core/emscripten/blob/main/emcc.py#L2435-L2440 introduced in https://github.com/emscripten-core/emscripten/pull/15905 ; when libxml2's configure script tries to build a trivial C program with `emcc`, the linker gets sad because it can't find the symbol `__funcs_on_exit`, which seems to be because the build flags include `-nostdlib` and no replacement is provided? I'm not clear on whether the misconfiguration is in libxml2 or in emscripten, though.

[config.log](https://github.com/NixOS/nixpkgs/files/8712119/config.log)

**Comment by duxovni** [comment]
https://github.com/emscripten-core/emscripten/pull/16986

File: pkgs/development/compilers/emscripten/default.nix [hunk 1]
```diff
@@ -38,6 +39,10 @@ stdenv.mkDerivation rec {
       src = ./0001-emulate-clang-sysroot-include-logic.patch;
       resourceDir = "${llvmEnv}/lib/clang/${llvmPackages.release_version}/";
     })
+    (fetchpatch {
+      url = "https://github.com/emscripten-core/emscripten/pull/16986.patch";
+      sha256 = "0wbnvdy6gyizmrvqwbdz9k2cs9cmni49iqn1hggsfcnkwn2qlddp";
```
Comment (willcohen) [inline pkgs/development/compilers/emscripten/default.nix:44]:
```suggestion
      # https://github.com/emscripten-core/emscripten/pull/16986
      url = "https://github.com/emscripten-core/emscripten/commit/d5ef6937fe395488e23a82c1e582a7ea5c2dab83.patch";
      sha256 = "sha256-YX5DG8i5x6S7XnU58etEapDd+o5SuzbFIGv8v/9+T3E=";
```
Might be cleaner to reference the actual commit that got merged upstream (also, the `sha` on the PR's patch was causing an error for me -- I had to run nix-build twice to get a working one), and then note the PR via a comment.

**Review by willcohen — CHANGES_REQUESTED** [review]

**Comment by duxovni** [comment]
(Could also move this to emscripten 3.1.11 since that includes my PR, but that's failing to build for possibly-LLVM-related reasons that I don't really have time to investigate further :/)

**Comment by cole-h** [comment]
@ofborg eval

**Review by willcohen — APPROVED** [review]

**Pull request merged**